### ISSUE TITLE
Token formats can also be advertised both by AS and RS

### ISSUE THREAD

**Issue by Denisthemalice** [ietf-wg-gnap/gnap-resource-servers#3]
Section 2. states:

> The core GNAP protocol makes no assumptions or demands on the format or contents of the access token, but such token formats can be the topic of agreements between the AS and RS.

Token formats can certainly be the topic of agreements between the AS and RS, but as presented this seems to be the single case.

It should be added that token formats can be advertised both by AS and RS, in a way where such token formats do not need to be the topic of agreements between the AS and RS. 

In this way a Client can contact both a RS and an AS and get sufficient information to negotiate the token format with the AS.

**Comment by fimbault** [comment]
The rationale here is that we're interested by the intersection of what the AS can generate and what the RS can understand. So it makes sense to focus on an agreement protocol. Otherwise we enter into potential incompatible sets, without a very clear reason as to why all that additional trial/error logic would be useful (to me at least).

**Comment by jricher** [comment]
The client is not part of this negotiation process between the AS and RS since the token format is opaque to the client, for privacy and security reasons as previously discussed in the WG.

**Comment by Denisthemalice** [comment]
**_fimbault_** wrote: 

> So it makes sense to focus on an agreement protocol. 

I fully agree.

**_@jricher_**: For privacy reasons, the token **_content_** (which is different from the token **_format_**) should **_NOT_** be opaque to the end user, but for implementation considerations, it MAY be simpler to consider the token content opaque to the client.

A **_token format_** is a header placed at the beginning of the token. Usually a "magic number" is used. Negotiating "magic numbers" between clients and both AS and RS does not mean that clients MUST necessarily understand the token **_content_**. 

The goal is simply to find an intersection between which token formats the AS can generate and which token formats the RS can understand, as **_fimbault_** noticed.

If this is done, it allows to overcome scalability issues, since the AS does not need to establish a prior agreement with every RS.
In terms of trust relationships (which unfortunately are still left undefined in both drafts), the RS needs to nominate which ASs it trusts (_and what for_) while the AS (as long as it does not work in relationship with one or more ROs) can be kept ignorant of the RS that are trusting it.  In this way, the RS does not need to contact the AS to identify and authenticate itself to the AS.

**Comment by aaronpk** [comment]
@Denisthemalice You are welcome to define an extension of GNAP that defines a token format and discovery mechanisms. There is precedent in this in the OAuth world already with the new JWT access token draft. However this is not going to be in the core document for all the reasons previously stated here and on the mailing list and during the interim meeting calls.

**Comment by Denisthemalice** [comment]
**_@jricher_**: You wrote:

> They are also certainly opaque to the user, as users are not software and have no business reading on-the-wire artifacts 
like access tokens. 

This argument does not stand. In order to access the content of an access token, an end-user needs a piece of software: its client. 
The goal is, for the end-user, to know whether the token contains some information that would allow different RSs to link their user accounts. If it does and if the end-user does not agree, the client should drop the access token.

At the moment, the presence of a "sub" claim (which is mandatory when using the access token format profile from OAuth) 
or of some equivalent is not discussed in the core document. This has important consequences about privacy. I opened an issue on that topic on the core document but this topic has been closed by **jricher** without providing any argument.

Security through opacity is not a good practice. In the same way, **_privacy through opacity_** is not acceptable 
... unless the (hidden) goal is have a system with built-in **Big Brothers**.

**_@aaronpk_**: we are far from being able to define an access token format since the discussion about the inclusion of roles, hierarchical group memberships, functional group memberships, attributes types with attribute values is still in its infancy.
